magpie/services.py

```diff
@@ -60,19 +68,82 @@ def resource_type_names(cls):
+             given resource considering allowed path structures.
+        """
         return len(cls.resource_types) > 0
 
 
 @six.add_metaclass(ServiceMeta)
 class ServiceInterface(object):
-    # required service type identifier (unique)
-    service_type = None                 # type: Str
-    # required request parameters for the service
-    params_expected = []                # type: List[Str]
-    # global permissions allowed for the service (top-level resource)
-    permissions = []                    # type: List[Permission]
-    # dict of list for each corresponding allowed resource permissions (children resources)
-    resource_types_permissions = {}     # type: Dict[models.Resource, List[Permission]]
+    @property
+    @abc.abstractmethod
+    def service_type(self):
+        # type: () -> Optional[Str]
+        """
+        Service type identifier (required, unique across implementation).
+        """
+        raise NotImplementedError
+
+    permissions = []  # type: List[Permission]
+    """
+    Permission allowed directly on the service as top-level resource.
+    """
+
+    resource_types_permissions = {}  # type: ResourceTypePermissions
+    """
+    Mapping of resource types to lists of permissions defining allowed children resource permissions under the service.
+    """
+
+    child_structure_allowed = []  # type: List[Str]
+    """
+    Control listing of path-like resource types limiting the allowed structure of nested children resources.
+
+    When not defined, any nested resource type combination is allowed if they themselves allow children resources.
+    Otherwise, nested child resource under the service can only be created at specific positions within the hierarchy
+    that matches exactly one of the listed control path-like definition. All definitions must start with ``service``
+    and must contain at least one separator and a sub-resource to ensure working behaviour of child resource under
+    the service.
+
+    For example, the below definition allows only resources typed ``route`` directly under the service.
+    The following nested resource under that first-level ``route`` can then be either another ``route`` followed
+    by a child ``process`` or directly a ``process``. Because ``process`` type doesn't allow any children resource
+    (see :attr:`models.Process.child_resource_allowed`), those are the only allowed combinations (cannot further nest
+    resources under the final ``process`` resource). Note that because intermediate ``route`` resources need to be
+    created at some point before the last ``process`` can even exist, partial paths (without ``process``) must also
+    be allowed as valid structures.
+
+    .. code-block:: python
+
+        child_structure_allowed = [
+            "service/route",
+            "service/route/process",
+            "service/route/route",
+            "service/route/route/process",
```

Comment:
The regex approach is exactly what I tried at first. ;) 
It looked like `service(/route)+(/process)?` to allow the route creation before `process` can exist. 

I encountered a problem though when trying to implement `resource_requested` because I somehow need to guess the "next type" from the path to provide which one are applicable during child resource creation, which became very complex with "any regex" form.

Reply:
Could it be simply resolved by defining the allowed child by resource type?
```
allowed_child_resource_type = {
   "service" : ["directory"],
   "directory" : ["directory", "file"],
   "file": []
}
allowed_child_resource_type = {
   "service" : ["route"],
   "route" : ["route", "process"],
   "process" : []
}
allowed_child_resource_type = {
   "service" : ["workspace"],
   "workspace" : ["layer"],
   "layer" : [] 
}
```
Or at some point you need to control the child type based on the current depth?

Reply:
It could work, except if we encounter a case were very specific nested placement is required. 
I'm thinking for example of  `service/route/route/route/leaf`, where there must be exactly 3 `route` items and only the last can have the `leaf` item.
Technically we could work around that by defining different classes for each of those items.

Reply:
Then, from my perspective, I find it neater to define a child relation, but if you want to merge this PR as is, I will not push for it.

Reply:
Given that I am currently creating the "paths" using format in the following manner, I tend to agree that using dicts like you propose would be more readable. 

```python 
"{}/{}/{}".format(
            models.Service.resource_type_name,
            models.Workspace.resource_type_name,
            models.Process.resource_type_name,
        )
```

I'll give it a try.

Reply:
Applied in 9c298f5

Reply:
Nice job! It looks like the config and the code handling it is cleaner.